providers/implementations/keymgmt/hybrid_kmgmt.c

```diff
@@ -0,0 +1,667 @@
+    OSSL_PARAM prms[2] = { OSSL_PARAM_END, OSSL_PARAM_END };
+    unsigned char *q;
+    char *str = NULL;
+    const OSSL_PARAM *p;
+    uint32_t key_length;
+    unsigned int i;
+    int type;
+
+    if (ossl_param_is_empty(params))
+        return 1;
+
+    for (p = params; params->key != NULL; params++) {
+        type = ossl_param_find_pidx(p->key);
+        switch (type) {
+        case PIDX_PKEY_PARAM_ENCODED_PUBLIC_KEY:
+            key_length = ossl_hybrid_get_key_length_from_key(key);
+            if (key_length != p->data_size) {
+                ERR_raise_data(ERR_LIB_PROV, PROV_R_INVALID_KEY_LENGTH,
+                               "length must be %u bytes, have %u bytes",
+                               key_length, p->data_size);
+                return 0;
+            }
+            q = p->data;
+            *prms = *p;
+            for (i = 0; i < key->num_algs; i++) {
+                prms->data = q;
+                prms->data_size = key->key_lengths[i];
+                q += key->key_lengths[i];
+                if (!EVP_PKEY_set_params(key->keys[i], prms))
+                    return 0;
+            }
+            break;
+
+        case PIDX_PKEY_PARAM_PROPERTIES:
+            if (!OSSL_PARAM_get_utf8_string(p, &str, SIZE_MAX))
+                return 0;
+            OPENSSL_free(key->propq);
+            key->propq = str;
+            /* Fall through */
+
+        default:
+            if (!key_subalg_set_param(key->num_algs, key->keys, prms))
+                return 0;
+            break;
+        }
+    }
+    return 1;
+}
+
+const OSSL_PARAM *ossl_hybrid_kmgmt_settable_params(void *vkey)
+{
+    HYBRID_PKEY *key = (HYBRID_PKEY *)vkey;
+    const OSSL_PARAM *pary[MAX_HYBRID_ALGS + 1];
+    unsigned int i;
+    static const OSSL_PARAM hybrid_settable_params[] = {
+        OSSL_PARAM_octet_string(OSSL_PKEY_PARAM_ENCODED_PUBLIC_KEY, NULL, 0),
+        OSSL_PARAM_utf8_string(OSSL_PKEY_PARAM_PROPERTIES, NULL, 0),
+        OSSL_PARAM_END
+    };
+
+    if (key->settable == NULL) {
+        for (i = 0; i < key->num_algs; i++)
+            pary[i] = EVP_PKEY_settable_params(key->keys[i]);
+        pary[i] = hybrid_settable_params;
+        key->settable = ossl_param_merge(key->num_algs + 1, pary);
+    }
+    return key->settable;
+}
+
+int ossl_hybrid_kmgmt_validate(const void *vctx, int selection, int checktype)
+{
+    HYBRID_PKEY_CTX *ctx = (HYBRID_PKEY_CTX *)vctx;
+    size_t key_length;
+    unsigned int i;
+    const int pub = (selection & OSSL_KEYMGMT_SELECT_PUBLIC_KEY) != 0;
+    const int priv = (selection & OSSL_KEYMGMT_SELECT_PRIVATE_KEY) != 0;
+
+    if (!ossl_prov_is_running())
+        return 0;
+    if ((key_length = ossl_hybrid_get_key_length_from_ctx(ctx)) == 0)
+        return 0;
+
+    if (pub || priv)
+        for (i = 0; i < ctx->num_algs; i++)
+            if ((priv && !EVP_PKEY_private_check(ctx->ctxs[i]))
+                    || (pub && !EVP_PKEY_public_check(ctx->ctxs[i])))
+                return 0;
```

Comment:
pairwise test also if they are both set?

Reply:
Not sure if fast/slow will apply to these?

Reply:
Good idea.